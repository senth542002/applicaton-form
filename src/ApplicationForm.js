import React, { Component } from 'react'
import DatePicker from 'react-datepicker'
import 'react-datepicker/dist/react-datepicker.css'

export default class ApplicationForm extends Component {
  constructor () {
    super()
    this.state = {
      student: {
        studentName: '',
        fatherName: '',
        motherName: '',
        email: '',
        mobileNumber: '',
        dateOfBirth: ''
      },
      successScreen: false,
      applicationNumber: '',
      isFormValid: false,
      errors: {}
    }

    this.submitFormHandler.bind(this)
    this.handleChange.bind(this)
  }

  handleValidation () {
    let student = this.state.student
    let errors = {}
    let formIsValid = true

    if (!student['studentName']) {
      formIsValid = false
      errors['studentName'] = 'Cannot be empty'
    }

    if (typeof student['studentName'] != 'undefined') {
      if (!student['studentName'].match(/^[a-zA-Z ]+$/)) {
        formIsValid = false
        errors['studentName'] = 'Only letters'
      }
    }

    if (!student['fatherName']) {
      formIsValid = false
      errors['fatherName'] = 'Cannot be empty'
    }

    if (typeof student['fatherName'] != 'undefined') {
      if (!student['fatherName'].match(/^[a-zA-Z ]+$/)) {
        formIsValid = false
        errors['fatherName'] = 'Only letters'
      }
    }

    if (!student['motherName']) {
      formIsValid = false
      errors['motherName'] = 'Cannot be empty'
    }

    if (typeof student['motherName'] != 'undefined') {
      if (!student['motherName'].match(/^[a-zA-Z ]+$/)) {
        formIsValid = false
        errors['motherName'] = 'Only letters'
      }
    }

    if (!student['email']) {
      formIsValid = false
      errors['email'] = 'Cannot be empty'
    }

    if (typeof student['email'] != 'undefined') {
      if (!student['email'].match(/^[a-zA-Z0-9]+@+[a-zA-Z0-9]+.+[A-z]/)) {
        formIsValid = false
        errors['email'] = 'Email is not valid'
      }
    }

    if (!student['mobileNumber']) {
      formIsValid = false
      errors['mobileNumber'] = 'Cannot be empty'
    }

    if (typeof student['mobileNumber'] != 'undefined') {
      if (!student['mobileNumber'].match(/^[0]?[789]\d{9}$/)) {
        formIsValid = false
        errors['mobileNumber'] = 'Mobile is not valid'
      }
    }

    if (!student['dateOfBirth']) {
      formIsValid = true
      errors['dateOfBirth'] = 'Cannot be empty'
    }

    this.setState({ errors: errors })
    return formIsValid
  }

  submitFormHandler = event => {
    event.preventDefault()
    let isFormValid = this.handleValidation()
    if (isFormValid) {
      console.log('Validaton Success')
      this.setState({
        successScreen: true,
        applicationNumber: Math.floor(100000 + Math.random() * 900000),
        isFormValid: isFormValid
      })
    } else {
      console.log('Validaton Errors')
    }
  }

  back = event => {
    event.preventDefault()
    this.setState({ successScreen: false })
  }

  handleChange = value => event => {
    if (value === 'dateOfBirth') {
      console.log(value + ' ' + event)
      // this.setState({
      //   ...this.state,
      //   [value] : event
      // })
      let student = this.state.student
      student[value] = event
      this.setState({ student })
    } else {
      console.log(value + ' ' + event.target.value)
      let student = this.state.student
      student[value] = event.target.value
      this.setState({ student })
    }
  }

  render () {
    return (
      <form className='demoForm'>
        <table className='responsive-table' hidden={this.state.successScreen}
        style={{ width: '500px'}}>
        <tbody>
            <tr
              className='form-group'
            >
              <td>
                <label className='studentName'>Student Name: </label>
              </td>
              <td>
                <input
                  type='text'
                  name='studentName'
                  value={this.state.student.studentName}
                  onChange={this.handleChange('studentName')}
                />
              </td>
              <td>
                <span style={{ color: 'red' }}>
                  {this.state.errors['studentName']}
                </span>
              </td>
            </tr>
            <tr
              className='form-group'
            >
              <td>
                <label className='fatherName'>Father Name: </label>
              </td>
              <td>
                <input
                  type='text'
                  name='fatherName'
                  value={this.state.student.fatherName}
                  onChange={this.handleChange('fatherName')}
                />
              </td>
              <td>
                <span style={{ color: 'red' }}>
                  {this.state.errors['fatherName']}
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <label className='motherName'>Mother Name: </label>
              </td>
              <td>
                <input
                  type='text'
                  name='motherName'
                  value={this.state.student.motherName}
                  onChange={this.handleChange('motherName')}
                />
              </td>
              <td>
                <span style={{ color: 'red' }}>
                  {this.state.errors['motherName']}
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <label className='email'>Email: </label>
              </td>
              <td>
                <input
                  type='email'
                  name='email'
                  value={this.state.student.email}
                  onChange={this.handleChange('email')}
                />
              </td>
              <td>
                <span style={{ color: 'red' }}>
                  {this.state.errors['email']}
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <label className='mobileNumber'>Mobile Number: </label>
              </td>
              <td>
                <input
                  type='text'
                  name='mobileNumber'
                  value={this.state.student.mobileNumber}
                  onChange={this.handleChange('mobileNumber')}
                />
              </td>
              <td>
                <span style={{ color: 'red' }}>
                  {this.state.errors['mobileNumber']}
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <label className='dateOfBirth'>Date of Birth: </label>
              </td>
              <td>
                <DatePicker
                  className='dateOfBirth'
                  name='dateOfBirth'
                  selected={this.state.student.dateOfBirth}
                  onChange={this.handleChange('dateOfBirth')}
                  dateFormat='MMMM d, yyyy'
                  timeCaption='time'
                />
              </td>
              <td>
                <span style={{ color: 'red' }}>
                  {this.state.errors['dateOfBirth']}
                </span>
              </td>
            </tr>
            </tbody>
            </table>

            <table className='responsive-table' style={{display: 'flex', alignItems: 'center', justifyContent: 'center'}}
            hidden={this.state.successScreen}>
            <tbody>
            <tr>
              <td>
                <button
                  className='submit'
                  type='Submit'
                  onClick={this.submitFormHandler}
                >
                  Submit
                </button>
              </td>
            </tr>
            </tbody>
        </table>
        <table hidden={!this.state.successScreen}>
          <tbody>
            <tr style={{ border: '3px solid grey', borderRightColor: 'grey' }}>
              <td>
                <label className='applicationNumber'>
                  Your application applicaton Number is :{' '}
                  {this.state.applicationNumber}
                </label>
              </td>
            </tr>
            <tr style={{ border: '3px solid grey', borderRightColor: 'grey' }}>
              <td>
                <label className='descripton'>
                  Application receipt will is sent to your email{' '}
                  {this.state.student.email}
                </label>
              </td>
            </tr>
            <tr
              style={{
                border: '3px solid grey',
                borderRightColor: 'grey',
                alignItems: 'center'
              }}
            >
              <td>
                <button className='back' type='Submit' onClick={this.back}>
                  Back
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    )
  }
}
