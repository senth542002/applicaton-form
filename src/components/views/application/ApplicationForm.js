import React, { Component } from 'react';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import applicationFormAPI from '../../api/ApplicationFormApi';
import pdfGeneratorAPI from '../../api/PdfGeneratorApi';
import LoadingOverlay from 'react-loading-overlay';
import BounceLoader from 'react-spinners/BounceLoader';
import FileDownload from 'js-file-download';
import { hashHistory, browserHistory, Link, withRouter  } from 'react-router';

export default class ApplicationForm extends Component {
  constructor () {
    super()
    this.state = {
      student: {
        name: '',
        fatherName: '',
        motherName: '',
        email: '',
        mobileNumber: '',
        dateOfBirth: ''
      },
      successScreen: false,
      applicationNumber: '',
      isFormValid: false,
      errors: {},
      active: false
    }

    this.submitFormHandler.bind(this)
    this.handleChange.bind(this)
  }

  componentDidMount() {
    browserHistory.push('/');
  }

  handleValidation () {
    console.log('inside handle Validation');
    let student = this.state.student
    let errors = {}
    let formIsValid = true

    if (!student['name']) {
      formIsValid = false
      errors['name'] = 'Cannot be empty'
    }

    if (typeof student['name'] != 'undefined') {
      if (!student['name'].match(/^[a-zA-Z ]+$/)) {
        formIsValid = false
        errors['name'] = 'Only letters'
      }
    }

    if (!student['fatherName']) {
      formIsValid = false
      errors['fatherName'] = 'Cannot be empty'
    }

    console.log("FatherName Validation:"+formIsValid);

    if (typeof student['fatherName'] != 'undefined') {
      if (!student['fatherName'].match(/^[a-zA-Z ]+$/)) {
        formIsValid = false
        errors['fatherName'] = 'Only letters'
      }
    }

    if (!student['motherName']) {
      formIsValid = false
      errors['motherName'] = 'Cannot be empty'
    }

    if (typeof student['motherName'] != 'undefined') {
      if (!student['motherName'].match(/^[a-zA-Z ]+$/)) {
        formIsValid = false
        errors['motherName'] = 'Only letters'
      }
    }

    if (!student['email']) {
      formIsValid = false
      errors['email'] = 'Cannot be empty'
    }

    if (typeof student['email'] != 'undefined') {
      if (!student['email'].match(/^[a-zA-Z0-9]+@+[a-zA-Z0-9]+.+[A-z]/)) {
        formIsValid = false
        errors['email'] = 'Email is not valid'
      }
    }

    if (!student['mobileNumber']) {
      formIsValid = false
      errors['mobileNumber'] = 'Cannot be empty'
    }

    if (typeof student['mobileNumber'] != 'undefined') {
      if (!student['mobileNumber'].match(/^[0]?[789]\d{9}$/)) {
        formIsValid = false
        errors['mobileNumber'] = 'Mobile is not valid'
      }
    }

    if (!student['dateOfBirth']) {
      formIsValid = false
      errors['dateOfBirth'] = 'Cannot be empty'
    }

    this.setState({ errors: errors })
    return formIsValid
  }

  submitFormHandler = event => {
    event.preventDefault();
    let isFormValid = this.handleValidation();
    if (isFormValid) {
      console.log('Validaton Success')
      this.setState({
        active: true
      })
      //applicationNumber: Math.floor(100000 + Math.random() * 900000),
      applicationFormAPI.post('/api/applications', this.state.student)
        .then(res => {
          console.log("Response:"+res)
          console.log("Id:"+res.data.id)

          //browserHistory.push('/success', {'applicationNumber': res.data.id});
          //  browserHistory.push({pathname: '/success', state: {applicationNumber: res.data.id}})
          this.setState({
            successScreen: false,
            applicationNumber: res.data.id,
            isFormValid: isFormValid,
            active: false
          })
          browserHistory.push({pathname: '/success', state: {applicationNumber: res.data.id, student: this.state.student }})
        })
        .catch(error => {
          console.log("Error:"+error)
          this.setState({
            active: false
          })
        })
    } else {
      console.log('Validaton Errors:')
    }
  }

  back = event => {
    event.preventDefault()
    this.setState({ successScreen: false })
  }

  handleChange = value => event => {
    if (value === 'dateOfBirth') {
      console.log(value + ' ' + event)
      let student = this.state.student
      student[value] = event
      this.setState({ student })
    } else {
      console.log(value + ' ' + event.target.value)
      let student = this.state.student
      student[value] = event.target.value
      this.setState({ student })
    }
  }

  viewFormHandler = event => {

    event.preventDefault();

      this.setState({
        active: true
      })
      console.log("before Call Pdf:"+this.state.student.name);

      pdfGeneratorAPI.post('/api/generate',this.state.student, { responseType: 'blob'})
        .then(res => {
           console.log("Response:" + res.data);
              this.setState({
                active: false
              })
            FileDownload(res.data, 'ApplicationForm.pdf');
        })
        .catch(error => {
          console.log("Error:"+error)
          this.setState({
            active: false
          })
        });
  }

  render () {
    return (

     <div className='App'>
           <header className='App-header'>
             <h2>Student Application Form</h2>
           </header>
           <div className='Application-header'>

           <form className='demoForm'>
                   <table className='responsive-table' hidden={this.state.successScreen}
                   style={{width: "110%"}}>
                   <tbody>
                       <tr
                         className='form-group'>
                         <td>
                           <label className='name'>Student Name: </label>
                         </td>
                         <td>
                           <input
                             type='text'
                             name='name'
                             value={this.state.student.name}
                             onChange={this.handleChange('name')}
                             style={{width: 300}}
                           />
                         </td>
                         <td>
                           <span style={{ color: 'red' }}>
                             {this.state.errors['name']}
                           </span>
                         </td>
                       </tr>
                       <tr
                         className='form-group'
                       >
                         <td>
                           <label className='fatherName'>Father Name: </label>
                         </td>
                         <td>
                           <input
                             type='text'
                             name='fatherName'
                             value={this.state.student.fatherName}
                             onChange={this.handleChange('fatherName')}
                             style={{width: 300}}
                           />
                         </td>
                         <td>
                           <span style={{ color: 'red' }}>
                             {this.state.errors['fatherName']}
                           </span>
                         </td>
                       </tr>
                       <tr>
                         <td>
                           <label className='motherName'>Mother Name: </label>
                         </td>
                         <td>
                           <input
                             type='text'
                             name='motherName'
                             value={this.state.student.motherName}
                             onChange={this.handleChange('motherName')}
                             style={{width: 300}}
                           />
                         </td>
                         <td>
                           <span style={{ color: 'red' }}>
                             {this.state.errors['motherName']}
                           </span>
                         </td>
                       </tr>
                       <tr>
                         <td>
                           <label className='email'>Email: </label>
                         </td>
                         <td>
                           <input
                             type='email'
                             name='email'
                             value={this.state.student.email}
                             onChange={this.handleChange('email')}
                             style={{width: 300}}
                           />
                         </td>
                         <td>
                           <span style={{ color: 'red' }}>
                             {this.state.errors['email']}
                           </span>
                         </td>
                       </tr>
                       <tr>
                         <td>
                           <label className='mobileNumber'>Mobile Number: </label>
                         </td>
                         <td>
                           <input
                             type='text'
                             name='mobileNumber'
                             value={this.state.student.mobileNumber}
                             onChange={this.handleChange('mobileNumber')}
                             style={{width: 300}}
                           />
                         </td>
                         <td>
                           <span style={{ color: 'red' }}>
                             {this.state.errors['mobileNumber']}
                           </span>
                         </td>
                       </tr>
                       <tr>
                         <td>
                           <label className='dateOfBirth'>Date of Birth: </label>
                         </td>
                         <td>
                           <DatePicker
                             className='dateOfBirth'
                             name='dateOfBirth'
                             selected={this.state.student.dateOfBirth}
                             onChange={this.handleChange('dateOfBirth')}
                             dateFormat='MMMM d, yyyy'
                             timeCaption='time'
                           />
                         </td>
                         <td>
                           <span style={{ color: 'red' }}>
                             {this.state.errors['dateOfBirth']}
                           </span>
                         </td>
                       </tr>
                       </tbody>
                       </table>
                       <div>
                         <LoadingOverlay active={this.state.active} spinner={<BounceLoader />} />
                       </div>
                       <table className='responsive-table' hidden={this.state.successScreen}
                       style={{display: 'flex', alignItems: 'center', justifyContent: 'center'}}
                       >
                       <tbody hidden={this.state.successScreen}>
                       <tr>
                         <td>
                           <button
                             className='submit'
                             type='Submit'
                             onClick={this.submitFormHandler}
                           >
                             Submit{this.state.successScreen}
                           </button>
                         </td>
                       </tr>
                       </tbody>
                   </table>
                   <table hidden={!this.state.successScreen}>
                     <tbody>
                       <tr style={{ border: '3px solid grey', borderRightColor: 'grey' }}>
                         <td>
                           <label className='applicationNumber'>
                             Your application applicaton Number is :{' '}
                             <a href= "#link" onClick={this.viewFormHandler} target="_blank">{this.state.applicationNumber}</a>
                           </label>
                         </td>
                       </tr>
                       <tr style={{ border: '3px solid grey', borderRightColor: 'grey' }}>
                         <td>
                           <label className='descripton'>
                             Application receipt will is sent to your email{' '}
                             {this.state.student.email}
                           </label>
                         </td>
                       </tr>
                       <tr
                         style={{
                           border: '3px solid grey',
                           borderRightColor: 'grey',
                           alignItems: 'center'
                         }}
                       >
                         <td>
                           <button className='back' type='Submit' onClick={this.back}>
                             Back
                           </button>
                         </td>
                       </tr>
                     </tbody>
                   </table>
                 </form>
           </div>
           <div className='rows' script='display:inline;' />
      </div>
    )
  }
}
