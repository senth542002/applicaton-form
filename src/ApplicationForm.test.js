import React from 'react'
import { shallow, mount } from 'enzyme'
import { renderer } from 'react-test-renderer'
import ApplicationForm from './ApplicationForm'

describe('ApplicationForm Component', () => {
  it('Verify Student Name', () => {
    const wrapper = shallow(<ApplicationForm />)
    expect(wrapper.exists('.studentName')).toEqual(true)
    //expect(text).toEqual('Student Name: ');
  })

  it('Captures studentname correctly on change', () => {
    const component = mount(<ApplicationForm />)
    const input = component.find('input').at(0)
    input.instance().value = 'Varun Karthik'
    input.simulate('change')
    expect(component.state().student.studentName).toEqual('Varun Karthik')
  })

  it('Navigates to success screen', () => {
    const component = mount(<ApplicationForm />)

    let input = component.find('input').at(0)
    input.instance().value = 'Varun Karthik'
    input.simulate('change')

    input = component.find('input').at(1)
    input.instance().value = 'Senthil Kumar'
    input.simulate('change')

    input = component.find('input').at(2)
    input.instance().value = 'Priya'
    input.simulate('change')

    input = component.find('input').at(3)
    input.instance().value = 'senth542002@gmail.com'
    input.simulate('change')

    input = component.find('input').at(4)
    input.instance().value = '9940206385'
    input.simulate('change')

    var student = component.state().student
    student['dateOfBirth'] = new Date('2019-07-08')
    component.setState({ student })
    const submitButton = component.find('button').at(0)
    submitButton.simulate('click')
    expect(component.state().successScreen).toEqual(true)
  })

  it('Validates form fields on Submit', () => {
    const component = mount(<ApplicationForm />)
    const submitButton = component.find('button').at(0)
    submitButton.simulate('click')

    const errors = {
      studentName: 'Only letters',
      fatherName: 'Only letters',
      motherName: 'Only letters',
      email: 'Email is not valid',
      mobileNumber: 'Mobile is not valid',
      dateOfBirth: 'Cannot be empty'
    }
    expect(component.state().errors).toEqual(errors)
  })
})
