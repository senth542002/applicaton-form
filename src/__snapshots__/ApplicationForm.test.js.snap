// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationForm Component matches the snapshot 1`] = `
<form
  className="demoForm"
>
  <table
    className="responsive-table"
    hidden={false}
  >
    <tbody>
      <tr
        className="form-group"
        tyle={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="studentName"
          >
            Student Name: 
          </label>
        </td>
        <td>
          <input
            name="studentName"
            onChange={[Function]}
            type="text"
            value=""
          />
        </td>
        <td>
          <span
            style={
              Object {
                "color": "red",
              }
            }
          />
        </td>
      </tr>
      <tr
        className="form-group"
        style={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="fatherName"
          >
            Father Name: 
          </label>
        </td>
        <td>
          <input
            name="fatherName"
            onChange={[Function]}
            type="text"
            value=""
          />
        </td>
        <td>
          <span
            style={
              Object {
                "color": "red",
              }
            }
          />
        </td>
      </tr>
      <tr
        style={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="motherName"
          >
            Mother Name: 
          </label>
        </td>
        <td>
          <input
            name="motherName"
            onChange={[Function]}
            type="text"
            value=""
          />
        </td>
        <td>
          <span
            style={
              Object {
                "color": "red",
              }
            }
          />
        </td>
      </tr>
      <tr
        style={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="email"
          >
            Email: 
          </label>
        </td>
        <td>
          <input
            name="email"
            onChange={[Function]}
            type="email"
            value=""
          />
        </td>
        <td>
          <span
            style={
              Object {
                "color": "red",
              }
            }
          />
        </td>
      </tr>
      <tr
        style={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="mobileNumber"
          >
            Mobile Number: 
          </label>
        </td>
        <td>
          <input
            name="mobileNumber"
            onChange={[Function]}
            type="text"
            value=""
          />
        </td>
        <td>
          <span
            style={
              Object {
                "color": "red",
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          <label
            className="dateOfBirth"
          >
            Date of Birth: 
          </label>
        </td>
        <td>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="dateOfBirth"
                disabled={false}
                name="dateOfBirth"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value=""
              />
            </div>
          </div>
        </td>
        <td>
          <span
            style={
              Object {
                "color": "red",
              }
            }
          />
        </td>
      </tr>
    </tbody>
    <tbody>
      <tr
        style={
          Object {
            "alignItems": "center",
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <button
            className="submit"
            onClick={[Function]}
            type="Submit"
          >
            Submit
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <table
    hidden={true}
  >
    <tbody>
      <tr
        style={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="applicationNumber"
          >
            Your application applicaton Number is :
             
            
          </label>
        </td>
      </tr>
      <tr
        style={
          Object {
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <label
            className="descripton"
          >
            Application receipt will is sent to your email
             
            
          </label>
        </td>
      </tr>
      <tr
        style={
          Object {
            "alignItems": "center",
            "border": "3px solid grey",
            "borderRightColor": "grey",
          }
        }
      >
        <td>
          <button
            className="back"
            onClick={[Function]}
            type="Submit"
          >
            Back
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</form>
`;
